homeassistant/components/keymitt_ble/config_flow.py

```diff
@@ -0,0 +1,164 @@
+        self, discovery_info: BluetoothServiceInfoBleak
+    ) -> FlowResult:
+        """Handle the bluetooth discovery step."""
+        _LOGGER.debug("Discovered bluetooth device: %s", discovery_info)
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        parsed = parse_advertisement_data(
+            discovery_info.device, discovery_info.advertisement
+        )
+        self._discovered_adv = parsed
+        data = parsed.data
+        self.context["title_placeholders"] = {
+            "name": data["local_name"],
+            "address": discovery_info.address,
+        }
+        return await self.async_step_init()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        # This is for backwards compatibility.
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Check if paired."""
+        errors: dict[str, str] = {}
+
+        if discovery := self._discovered_adv:
+            self._discovered_advs[discovery.address] = discovery
+            self._name = name_from_discovery(discovery)
+        else:
+            self._name = DEFAULT_NAME
+            current_addresses = self._async_current_ids()
+            for discovery_info in async_discovered_service_info(self.hass):
+                address = discovery_info.address
+                if (
+                    address in current_addresses
+                    or address in self._discovered_advs
+                ):
+                    continue
+                parsed = parse_advertisement_data(
+                    discovery_info.device, discovery_info.advertisement
+                )
+                if parsed:
+                    self._discovered_advs[address] = parsed
+
+        if not self._discovered_advs:
+            return self.async_abort(reason="no_unconfigured_devices")
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_BDADDR): vol.In(
+                    {
+                        address: f"{parsed.data['local_name']} ({address})"
+                        for address, parsed in self._discovered_advs.items()
+                    }
+                )
+            }
+        )
+
+        if user_input is not None:
+            self._bdaddr = user_input[CONF_BDADDR]
+            await self.async_set_unique_id(
+                self._bdaddr, raise_on_progress=False
+            )
+            self._abort_if_unique_id_configured()
+            self._ble_device = bluetooth.async_ble_device_from_address(
+                self.hass, self._bdaddr
+            )
+            if not self._ble_device:
+                raise IntegrationError(
+                    f"Could not find MicroBot with address {self._bdaddr}"
+                )
```

Comment:
Grab the device from `self._discovered_advs[address].device` so this can't happen